drivers/SmartThings/zigbee-lock/src/test/test_zigbee_lock_code_slga_migration.lua

```diff
@@ -0,0 +1,92 @@
+
+local clusters = require "st.zigbee.zcl.clusters"
+local PowerConfiguration = clusters.PowerConfiguration
+local DoorLock = clusters.DoorLock
+local Alarm = clusters.Alarms
+local capabilities = require "st.capabilities"
+
+local json = require "st.json"
+
+local mock_datastore = require "integration_test.mock_env_datastore"
+
+local mock_device = test.mock_device.build_test_zigbee_device(
+    {
+      profile = t_utils.get_profile_definition("base-lock.yml"),
+      data = {
+        lockCodes = json.encode({
+          ["1"] = "Zach",
+          ["2"] = "Steven"
+        }),
+      }
+    }
+)
+
+zigbee_test_utils.prepare_zigbee_env_info()
+local function test_init()end
+
+test.set_test_init_function(test_init)
+
+test.register_coroutine_test(
+    "Device called 'migrate' command",
+    function()
+        test.mock_device.add_test_device(mock_device)
+        test.socket.device_lifecycle:__queue_receive({ mock_device.id, "added" })
+        test.socket.zigbee:__expect_send({ mock_device.id, PowerConfiguration.attributes.BatteryPercentageRemaining:read(mock_device) })
+        test.socket.zigbee:__expect_send({ mock_device.id, DoorLock.attributes.LockState:read(mock_device) })
+        test.socket.zigbee:__expect_send({ mock_device.id, Alarm.attributes.AlarmCount:read(mock_device) })
+        test.wait_for_events()
+        -- Validate lockCodes field
+        mock_datastore.__assert_device_store_contains(mock_device.id, "lockCodes", { ["1"] = "Zach", ["2"] = "Steven" })
+        -- Validate state cache
+        mock_datastore.__assert_device_store_contains(mock_device.id, "__state_cache",
+            {
+                main = {
+                lockCodes = {
+                    lockCodes = {value = json.encode({ ["1"] = "Zach", ["2"] = "Steven" }) }
+                }
+                }
+            }
+        )
+        -- Validate migration complete flag
+        mock_datastore.__assert_device_store_contains(mock_device.id, "migrationComplete", true)
+
+        -- Set min/max code length attributes
+        test.socket.zigbee:__queue_receive({ mock_device.id, DoorLock.attributes.MinPINCodeLength:build_test_attr_report(mock_device, 5) })
+        test.socket.zigbee:__queue_receive({ mock_device.id, DoorLock.attributes.MaxPINCodeLength:build_test_attr_report(mock_device, 10) })
+        test.socket.zigbee:__queue_receive({ mock_device.id, DoorLock.attributes.NumberOfPINUsersSupported:build_test_attr_report(mock_device, 4) })
+        test.socket.capability:__expect_send( mock_device:generate_test_message("main", capabilities.lockCodes.minCodeLength(5,  { visibility = { displayed = false } })))
+        test.socket.capability:__expect_send( mock_device:generate_test_message("main", capabilities.lockCodes.maxCodeLength(10,  { visibility = { displayed = false } })))
+        test.socket.capability:__expect_send( mock_device:generate_test_message("main", capabilities.lockCodes.maxCodes(4,  { visibility = { displayed = false } })))
+
```

Comment:
I think you want a `test.wait_for_events()` here. I was getting occasional failures without it.